Updated PyChiquito frontend for latest changes

qwang98 opened the pull request:
Most specifically, need to acquire num_step from ast to calculate padding in PyChiquito.

leolara commented:
@qwang98 I don't understand, the num_steps is set in the python frontend, and the padding is done in the python frontend. So I don't understand what you mean about getting num_steps from AST. The AST is in the python frontend where the padding is done.

qwang98 commented:
Sorry for the confusion. My description was incorrect and doesn't match the changes. The changes are:
1. Store AST in the CIRCUIT_MAP, so that the MockProver can obtain whether there's exposed columns in the AST, and fill in the third `instance` parameter of `MockProver::run` accordingly, i.e. empty Vec if there's no instance or non-empty Vec if there is.
3. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
4. Added empty trace function for AST Circuit deserializer, because otherwise the compiler won't generate witness if trace defaults to None.

@leolara Ready for review again :)

leolara commented:
> 1. Store AST in the CIRCUIT_MAP, so that the MockProver can obtain whether there's exposed columns in the AST, and fill in the third `instance` parameter of `MockProver::run` accordingly, i.e. empty Vec if there's no instance or non-empty Vec if there is.

So the solution for this is explained here: https://github.com/privacy-scaling-explorations/chiquito/issues/83 and also the exposed column is in the plonkish IR and in the ChiquitoHalo2 so it wouldn't be necessary to keep the AST for that.

leolara commented:
> 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.

Why is there plaf stufff in the python frontend? I think for now we should not implement plaf backend for the python frontend

leolara commented:
> 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.

I think this is better solved by implementing Debug and/or Display traits on certain types.

qwang98 commented:
> > 1. Store AST in the CIRCUIT_MAP, so that the MockProver can obtain whether there's exposed columns in the AST, and fill in the third `instance` parameter of `MockProver::run` accordingly, i.e. empty Vec if there's no instance or non-empty Vec if there is.
> 
> So the solution for this is explained here: #83 and also the exposed column is in the plonkish IR and in the ChiquitoHalo2 so it wouldn't be necessary to keep the AST for that.

Cool. Working on this now :)

qwang98 commented:
> > 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
> 
> Why is there plaf stufff in the python frontend? I think for now we should not implement plaf backend for the python frontend

It's not implementing Plaf backend. It's really just printing witness assignment with the help of Plaf.

qwang98 commented:
> > 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
> 
> I think this is better solved by implementing Debug and/or Display traits on certain types.

I see your point now. I'll think about it and either implement this or remove the print_plaf_witness function entirely. :)

leolara commented:
The optimal solution would be to implement unit tests

On Thu, 17 Aug 2023 at 11:43, Steve Wang ***@***.***> wrote:

>
>    1. Added a debug function that prints witness after converting it to
>    Plaf. This was quite helpful for identifying a couple bugs that I fixed. I
>    exposed it in PyO3 but not in Python DSL, as it's not some function the
>    user will use.
>
> I think this is better solved by implementing Debug and/or Display traits
> on certain types.
>
> I see your point now. I'll think about it and either implement this or
> remove the print_plaf_witness function entirely. :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/privacy-scaling-explorations/chiquito/pull/84#issuecomment-1681604571>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJHQ4UQHOXZPNNH25M5X4DXVWOQLANCNFSM6AAAAAA3SJHXME>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

leolara commented:
> > > 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
> > 
> > 
> > Why is there plaf stufff in the python frontend? I think for now we should not implement plaf backend for the python frontend
> 
> It's not implementing Plaf backend. It's really just printing witness assignment with the help of Plaf.

We should implement the Debug and Display for TraceWitness struct, Assigments, etc... much easier that way. But can be done in another PR

leolara commented:
@qwang98 I think we can make it for the super circuit, if we follow the same order as the meta.instance_columns are called.

leolara commented on src/plonkish/backend/halo2.rs:
```diff
@@ -388,13 +415,35 @@ pub struct ChiquitoHalo2SuperCircuit<F: Field + From<u64>> {
+
+    pub fn instance(&self) -> Vec<Vec<F>> {
+        if !self.witness.is_empty() {
+            let mut result = Vec::new();
```
if you put let mut result outside if, you don't need the else. Just write result in the last line

qwang98 replied:
Indeed.

leolara commented on src/plonkish/backend/halo2.rs:
```diff
@@ -388,13 +415,35 @@ pub struct ChiquitoHalo2SuperCircuit<F: Field + From<u64>> {
+        if !self.witness.is_empty() {
+            let mut result = Vec::new();
+
+            for (uuid, assignment) in &self.witness {
```
@qwang98 I don't think you should follow the order of the witness, but the order of sub_circuit, or whatever order is used to call meta.instance_column()

qwang98 replied:
Agreed. sub_circuit is the order meta.instance_column is called.

qwang98 commented:
> @qwang98 I think we can make it for the super circuit, if we follow the same order as the meta.instance_columns are called.

Done

qwang98 commented:
> > > > 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
> > > 
> > > 
> > > Why is there plaf stufff in the python frontend? I think for now we should not implement plaf backend for the python frontend
> > 
> > 
> > It's not implementing Plaf backend. It's really just printing witness assignment with the help of Plaf.
> 
> We should implement the Debug and Display for TraceWitness struct, Assigments, etc... much easier that way. But can be done in another PR

Agree on this.
